src/System/reachability.rs

```diff
@@ -0,0 +1,199 @@
+use edbm::zones::OwnedFederation;
+
+use crate::ModelObjects::component::{State, Transition};
+use crate::TransitionSystems::{LocationID, TransitionSystem};
+use std::collections::HashMap;
+use std::rc::Rc;
+
+/// This hold the result of a reachability query
+pub struct Path {
+    pub path: Option<Vec<Transition>>,
+    pub was_reachable: bool,
+}
+
+struct SubPath {
+    previous_sub_path: Option<Rc<SubPath>>,
+    destination_state: State,
+    transition: Option<Transition>,
+}
+
+fn is_trivially_unreachable(end_state: &State, _system: &dyn TransitionSystem) -> bool {
+    if let Some(invariants) = end_state.get_location().get_invariants() {
+        if !&end_state.zone_ref().has_intersection(invariants) {
+            return true;
+        }
+    }
+    false
+}
+///# Find path
+///
+/// Returns a path from a start state to an end state in a transition system.
+///
+/// If it is reachable, it returns a path.
+///
+/// If it is not reachable, it returns None.
+///
+/// The start state can be omitted with None to use the start state of the transition system.
+///
+///## Checking if a state can reach another:
+/// ```ignore
+/// let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
+///    Ok(result) => match result {
+///        Some(path) => true,
+///        None => false,
+///    },
+///    Err(string) => panic!(string),
+/// };
+/// ```
+///
+///## Omitting start state:
+/// ```ignore
+/// let is_reachable: bool = match find_path(None, end_state, transition_system) {
+///    Ok(result) => match result {
+///        Some(path) => true,
+///        None => false,
+///    },
+///    Err(string) => panic!(string),
+/// };
+/// ```
+pub fn find_path(
+    start_state: State,
+    end_state: State,
+    system: &dyn TransitionSystem,
+) -> Result<Path, String> {
+    if is_trivially_unreachable(&end_state, system) {
+        return Ok(Path {
+            path: None,
+            was_reachable: false,
+        });
+    }
+
+    Ok(search_algorithm(&start_state, &end_state, system))
+}
+
+fn search_algorithm(start_state: &State, end_state: &State, system: &dyn TransitionSystem) -> Path {
+    // Apply the invariant of the start state to the start state
+    let mut start_clone = start_state.clone();
+    let start_zone = start_clone.take_zone();
+    let zone = start_clone.decorated_locations.apply_invariants(start_zone);
+    start_clone.set_zone(zone);
+
+    // hashmap linking every location to all its current zones
+    let mut visited_states: HashMap<LocationID, Vec<OwnedFederation>> = HashMap::new();
+    //let mut made_transitions: Vec<SubPath> = Vec::new();
+
+    // List of states that are to be visited
+    let mut frontier_states: Vec<Rc<SubPath>> = Vec::new();
+
+    let mut actions: Vec<String> = system.get_actions().into_iter().collect();
+    actions.sort();
```

Comment:
Do we need to sort the actions?

Reply:
Da det er et hashset , så vi er nødt til at have samme orden hver gang så fx den samme query, så mister vi vores test så vi ikke kan finde de mange forskellige kombinationer.